src/hotspot/share/opto/loopnode.cpp

```diff
@@ -6058,6 +6058,24 @@ CountedLoopEndNode* CountedLoopNode::find_pre_loop_end() {
   return pre_end;
 }
 
+Node* CountedLoopNode::uncasted_init_trip(bool uncast) {
+  Node* init = init_trip();
+  if (uncast && init->is_CastII()) {
+    // skip over the cast added by PhaseIdealLoop::cast_incr_before_loop() when pre/post/main loops are created because
+    // it can get in the way of type propagation. For instance, the index tested by an assert predicate, if the cast is
+    // not skipped over, could be (1):
+    // (AddI (CastII (AddI pre_loop_iv -2) int) 1)
+    // while without the cast, it is (2):
+    // (AddI (AddI pre_loop_iv -2) 1)
+    // which is be transformed to (3):
+    // (AddI pre_loop_iv -1)
+    // The compiler may be able to constant fold the assert predicate condition for (3) but not (1)
```

Comment:
```suggestion
    // skip over the cast added by PhaseIdealLoop::cast_incr_before_loop() when pre/post/main loops are created because
    // it can get in the way of type propagation. For instance, the index tested by an Assertion Predicate, if the cast is
    // not skipped over, could be (1):
    // (AddI (CastII (AddI pre_loop_iv -2) int) 1)
    // while without the cast, it is (2):
    // (AddI (AddI pre_loop_iv -2) 1)
    // which is be transformed to (3):
    // (AddI pre_loop_iv -1)
    // The compiler may be able to constant fold the Assertion Predicate condition for (3) but not (1)
```